File: solution/35.search-insert-position/binary-search.rs
/*
 * @lc app=leetcode.cn id=35 lang=rust
 *
 * https://leetcode-cn.com/problems/search-insert-position
 */

// @lc code=start
impl Solution {
    pub fn search_insert(nums: Vec<i32>, target: i32) -> i32 {
        let mut left = 0;
        let mut right = nums.len() as i32 - 1;
        while left <= right {
            let middle = (right - left) / 2 + left;
            let value = nums[middle as usize];
            if value == target {
                return middle;
            } else if value > target {
                right = middle - 1;
            } else {
                left = middle + 1;
            }
        }
        left
    }
}
// @lc code=end

File: solutions/33.search-in-rotated-sorted-array/src/binary_search.ts
/*
 * @lc app=leetcode.cn id=33 lang=typescript
 *
 * https://leetcode.cn/problems/search-in-rotated-sorted-array/
 */

// @lc code=start
function search(nums: number[], target: number): number {
    let left = 0;
    let right = nums.length - 1;
    while (left <= right) {
        const mid = (left + right) >> 1;
        if (nums[mid] === target) {
            return mid;
        }
        if (nums[left] <= nums[mid]) {
            if (nums[left] <= target && target < nums[mid]) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        } else {
            if (nums[mid] < target && target <= nums[right]) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
    }
    return -1;
}
// @lc code=end

export { search };
